content/blogposts/2022/multi-version-upgrades.md

```diff
@@ -0,0 +1,109 @@
+
+In an unrelated change, we [explicitly mark privileged migrations](https://github.com/sourcegraph/sourcegraph/pull/31782). Privileged migrations contain certain operations that require elevated or superuser privileges, such as enabling a Postgres extension. Marking these migrations statically allow us to present reduced-privileged users with the required manual steps to perform during the migration process.
+
+In order to keep track of the various migration metadata, we [reorganized migration definitions into directories](https://github.com/sourcegraph/sourcegraph/pull/30276), each with three files: `up.sql`, `down.sql`, and `metadata.yaml` file, as shown in the following image. Also of particular interest is the introduction of two entirely new and distinct database schemas controlled by the same migration infrastructure ([codeintel](https://github.com/sourcegraph/sourcegraph/pull/13903) and [codeinsights](https://github.com/sourcegraph/sourcegraph/pull/17432)).
+
+<Figure src="https://storage.googleapis.com/sourcegraph-assets/blog/multi-version-upgrades/tree.png" alt="File tree view of the (updated) migrations directory." />
+
+In what is likely the largest departure from golang-migrate, we allow migrations to [define multiple parents](https://github.com/sourcegraph/sourcegraph/pull/30664). This allows our migration definitions to form a [directed acyclic graph](https://en.wikipedia.org/wiki/Directed_acyclic_graph) (instead of an absolutely ordered list). Note that in the image above our migrations no longer form a gapless sequence, and a [topological sort](https://en.wikipedia.org/wiki/Topological_sorting) over the migration graph is done to determine an application order.
+
+Well that seemed like a bunch of work for no payoff. **Why do that at all?** Short answer: engineering ergonomics. When we had a gapless sequence we would **frequently** run into merge conflicts when two engineers add migrations around the same time. Since each branch did not contain the other's migration, they would both choose the same sequence number. Because of limitations in our CI (no build queues + no non-stale base branch merge requirement), this conflict would, the majority of the time, be caught only after the both branchs have been merged and validation errors turned the `main` branch to go red.
```

Comment:
```suggestion
Well that seemed like a bunch of work for no payoff. **Why do that at all?** Short answer: engineering ergonomics. When we had a gapless sequence we would **frequently** run into merge conflicts when two engineers add migrations around the same time. Since each branch did not contain the other's migration, they would both choose the same sequence number. Because of limitations in our CI (no build queues + no non-stale base branch merge requirement), this conflict would, the majority of the time, be caught only after both branches have been merged and validation errors turned the `main` branch to go red.
```